runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -359,16 +359,35 @@ end
 ---@param bufnr number
 ---@param client_id number|nil If nil, then return all of the diagnostics.
 ---                            Else, return just the diagnostics associated with the client_id.
-function M.get(bufnr, client_id)
+---@param opts table|nil Configuration keys
+---         - severity: (DiagnosticSeverity, default nil)
+---             - Only return diagnostics with this severity. Overrides severity_limit
+---         - severity_limit: (DiagnosticSeverity, default nil)
+---             - Limit severity of diagnostics found. E.g. "Warning" means { "Error", "Warning" } will be valid.
+---@param predicate function|nil Optional function for filtering diagnostics
+---@param comp      function|nil Optional function for sorting diagnostics
+function M.get(bufnr, client_id, opts, predicate, comp)
+  opts = opts or {}
+  predicate = predicate or function(_) return true end
   if client_id == nil then
     local all_diagnostics = {}
     for iter_client_id, _ in pairs(diagnostic_cache[bufnr]) do
       local iter_diagnostics = M.get(bufnr, iter_client_id)
 
       for _, diagnostic in ipairs(iter_diagnostics) do
-        table.insert(all_diagnostics, diagnostic)
+        if predicate(diagnostic) then
+          table.insert(all_diagnostics, diagnostic)
+        end
       end
     end
+    if opts.severity then
+      all_diagnostics = filter_to_severity_limit(opts.severity, all_diagnostics)
+    elseif opts.severity_limit then
+      all_diagnostics = filter_by_severity_limit(opts.severity_limit, all_diagnostics)
+    end
+    if comp then
+      table.sort(all_diagnostics, comp)
+    end
 
     return all_diagnostics
   end
```

Comment:
It would also be necessary to apply the predicate to the contents of `diagnostic_cache` further below